# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [1]:
import pandas as pd

In [4]:
url = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv'
customer_df_all = pd.read_csv(url)
customer_df_all

Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     Number of Open Complaints     Policy Type  Vehicle Class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      Total Claim Amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

In [5]:
customer_df = customer_df_all.dropna(how="all")
customer_df

Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
1066  TM65736      Oregon      M                Master   
1067  VJ51327        Cali      F  High School or Below   
1068  GS98873     Arizona      F              Bachelor   
1069  CW49887  California      F                Master   
1070  MY31220  California      F               College   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
1066              305955.03%  38644.0                  78.0   
1067             2031499.76%  63209.0                 102.0   
1068              323912.47%  16061.0                  88.0   
1069              462680.11%  79487.0                 114.0   
1070              899704.02%  54230.0                 112.0   

     Number of Open Complaints     Policy Type  Vehicle Class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
1066                    1/1/00   Personal Auto  Four-Door Car   
1067                    1/2/00   Personal Auto            SUV   
1068                    1/0/00   Personal Auto  Four-Door Car   
1069                    1/0/00    Special Auto            SUV   
1070                    1/0/00   Personal Auto   Two-Door Car   

      Total Claim Amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
1066          361.455219  
1067          207.320041  
1068          633.600000  
1069          547.200000  
1070          537.600000  

[1071 rows x 11 columns]

In [70]:
customer_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1071 entries, 0 to 1070
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   1071 non-null   object 
 1   ST                         1071 non-null   object 
 2   GENDER                     954 non-null    object 
 3   Education                  1071 non-null   object 
 4   Customer Lifetime Value    1068 non-null   object 
 5   Income                     1071 non-null   float64
 6   Monthly Premium Auto       1071 non-null   float64
 7   Number of Open Complaints  1071 non-null   object 
 8   Policy Type                1071 non-null   object 
 9   Vehicle Class              1071 non-null   object 
 10  Total Claim Amount         1071 non-null   float64
dtypes: float64(3), object(8)
memory usage: 132.7+ KB


In [8]:
print("Dimensions original df: ", customer_df_all.shape)
print("Dimensions without NAN rows: ", customer_df.shape)


print()
print(customer_df.describe())

print()
print(customer_df.describe(include="object"))



Dimensions original df:  (4008, 11)
Dimensions without NAN rows:  (1071, 11)

             Income  Monthly Premium Auto  Total Claim Amount
count   1071.000000           1071.000000         1071.000000
mean   39295.701214            193.234360          404.986909
std    30469.427060           1601.190369          293.027260
min        0.000000             61.000000            0.382107
25%    14072.000000             68.000000          202.157702
50%    36234.000000             83.000000          354.729129
75%    64631.000000            109.500000          532.800000
max    99960.000000          35354.000000         2893.239678

       Customer      ST GENDER Education Customer Lifetime Value  \
count      1071    1071    954      1071                    1068   
unique     1071       8      5         6                    1027   
top     RB50392  Oregon      F  Bachelor              445811.34%   
freq          1     320    457       324                       4   

       Number of Open 

In [9]:
print()
print("Info: ", customer_df.info())


<class 'pandas.core.frame.DataFrame'>
Index: 1071 entries, 0 to 1070
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   1071 non-null   object 
 1   ST                         1071 non-null   object 
 2   GENDER                     954 non-null    object 
 3   Education                  1071 non-null   object 
 4   Customer Lifetime Value    1068 non-null   object 
 5   Income                     1071 non-null   float64
 6   Monthly Premium Auto       1071 non-null   float64
 7   Number of Open Complaints  1071 non-null   object 
 8   Policy Type                1071 non-null   object 
 9   Vehicle Class              1071 non-null   object 
 10  Total Claim Amount         1071 non-null   float64
dtypes: float64(3), object(8)
memory usage: 100.4+ KB
Info:  None


## Observations:
### Dimension:
- Ca. 3000 rows have all NAN value -> only 1071 rows have at least one value.
### Data Types:
- Customer Lifetime Value would better be a float.
- Total Claim Amount could have 2 decimals for better readability.

In [10]:
for col in customer_df.columns:
    if col.lower() != "income" and col.lower() != "monthly premium auto" and col.lower() != "total claim amount":
        print(col, customer_df[col].nunique(), list(customer_df[col].unique()))

Customer 1071 ['RB50392', 'QZ44356', 'AI49188', 'WW63253', 'GA49547', 'OC83172', 'XZ87318', 'CF85061', 'DY87989', 'BQ94931', 'SX51350', 'VQ65197', 'DP39365', 'SJ95423', 'IL66569', 'BW63560', 'FV94802', 'OE15005', 'WC83389', 'FL50705', 'ZK25313', 'QK46697', 'YH23384', 'TZ98966', 'HM55802', 'FS42516', 'US89481', 'HS14476', 'GE62437', 'EJ77678', 'SV85652', 'UL64533', 'PF41800', 'HD95276', 'SK67821', 'YV55495', 'KY38074', 'DM79012', 'CM61827', 'WC35801', 'QG25316', 'MB98372', 'IL19217', 'SR38658', 'YD87931', 'HG65722', 'BU27331', 'XM45289', 'KP34198', 'SH90947', 'WE95729', 'PY51963', 'RB69909', 'NW21079', 'FR46645', 'SY17488', 'AP67935', 'FS37417', 'ML29312', 'UB61619', 'CD86811', 'RU83859', 'FG63582', 'NN71951', 'WB37082', 'SM52139', 'FL82372', 'DP45816', 'GW33762', 'RZ33670', 'PY70169', 'MO91628', 'HW87852', 'HB20453', 'BN87372', 'YX23800', 'DZ87709', 'XW13033', 'SP81997', 'OM82309', 'ZU35962', 'VH85817', 'DT85712', 'YJ88573', 'SQ19467', 'YB66933', 'ET79815', 'QC35222', 'CJ15590', 'OI482

In [11]:
customer_df["ST"].value_counts()

ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64

In [12]:
customer_df["Total Claim Amount"].describe()

count    1071.000000
mean      404.986909
std       293.027260
min         0.382107
25%       202.157702
50%       354.729129
75%       532.800000
max      2893.239678
Name: Total Claim Amount, dtype: float64

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [30]:
location = customer_df["ST"]
# print("Locations frequency: ", location.value_counts())
less_common_locations = location.value_counts().sort_values(ascending=[True])[:5]
print("Less common locations: ", less_common_locations)

Less common locations:  ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: count, dtype: int64


### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [32]:
customer_df["Policy Type"].value_counts()

Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [48]:
policy_index = customer_df["Policy Type"].values

print("Average Income Personal Auto Policy Holder:" , round(customer_df.loc[customer_df['Policy Type'] == 'Personal Auto']["Income"].mean(), 2))
print("Average Income Corporate Auto Policy Holder:" , round(customer_df.loc[customer_df['Policy Type'] == 'Corporate Auto']["Income"].mean(), 2))

Average Income Personal Auto Policy Holder: 38180.7
Average Income Corporate Auto Policy Holder: 41390.31


### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [66]:
claim_percentile_75 = customer_df["Total Claim Amount"].quantile(0.75)

high_claimers = customer_df[customer_df["Total Claim Amount"] > claim_percentile_75].sort_values(by="Total Claim Amount", ascending=False)

high_claimers

Customer          ST GENDER             Education  \
888   BZ65376     Arizona      M              Bachelor   
248   FO35655      Oregon      M              Bachelor   
522   ZZ97035  California      M               College   
196   AZ62651      Oregon      M  High School or Below   
581   PR53785     Arizona      F  High School or Below   
...       ...         ...    ...                   ...   
590   UK41984     Arizona      F                Doctor   
804   QC89139      Oregon      F  High School or Below   
66    FL82372      Oregon    NaN               College   
218   KY33386          WA  Femal  High School or Below   
1070  MY31220  California      F               College   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
888               858127.87%  27689.0                 239.0   
248               867222.97%      0.0                 245.0   
522              2071494.04%      0.0                 203.0   
196               995170.77%  56855.0                 255.0   
581               745723.78%      0.0                 198.0   
...                      ...      ...                   ...   
590               383960.61%      0.0                 112.0   
804               452873.74%  90034.0                 112.0   
66                867219.43%  22547.0                 112.0   
218               800739.94%      0.0                 112.0   
1070              899704.02%  54230.0                 112.0   

     Number of Open Complaints     Policy Type Vehicle Class  \
888                     1/2/00   Personal Auto    Luxury SUV   
248                     1/0/00  Corporate Auto    Luxury SUV   
522                     1/0/00  Corporate Auto    Luxury Car   
196                     1/0/00  Corporate Auto    Luxury SUV   
581                     1/1/00   Personal Auto    Luxury Car   
...                        ...             ...           ...   
590                     1/2/00   Personal Auto           SUV   
804                     1/0/00  Corporate Auto           SUV   
66                      1/0/00  Corporate Auto           SUV   
218                     1/0/00   Personal Auto    Sports Car   
1070                    1/0/00   Personal Auto  Two-Door Car   

      Total Claim Amount  
888          2893.239678  
248          2345.413441  
522          2027.724442  
196          1836.000000  
581          1577.674417  
...                  ...  
590           537.600000  
804           537.600000  
66            537.600000  
218           537.600000  
1070          537.600000  

[264 rows x 11 columns]

In [71]:
high_claimers["Total Claim Amount"].describe().apply("{0:.2f}".format)

count     264.00
mean      782.23
std       292.75
min       537.60
25%       606.52
50%       679.60
75%       851.40
max      2893.24
Name: Total Claim Amount, dtype: object